File: Code/C++/study.cpp
#include "study.h"

using namespace std;

/*Constructeur de base*/
Study :: Study(
    int size, 
    int equiSample,
    int meanSample,
    double tempStart,
    double tempEnd,
    double tempSample,
    bool startRandom,
    string outputFile) :
    montecarlo(size, equiSample, meanSample, tempStart, startRandom),
    compteur(0),
    tempSample(tempSample),
    outputFile(outputFile) {

    Temperatures.resize(tempSample);
    Energies.resize(tempSample);
    Orders.resize(tempSample);

    for (int t = 0; t < tempSample; ++t) {
        Temperatures[t] = tempStart + (tempEnd - tempStart)*t/tempSample;
    }
}

/*Constructeur de base*/
Study :: Study(string basename) :
    montecarlo(basename) {

    string filename = basename + "_study.save";
    ifstream flux(filename.c_str(), ios::in);

    if(flux) {
        flux >> outputFile;
        flux >> tempSample;
        flux >> compteur;

        Temperatures.resize(tempSample);
        Energies.resize(tempSample);
        Orders.resize(tempSample);
        for (int t = 0; t < tempSample; ++t) {
                    flux >> Temperatures[t];
                    flux >> Energies[t][0];
                    flux >> Energies[t][1];
                    flux >> Orders[t][0];
                    flux >> Orders[t][1];
        }
        cout << BOLDGREEN << "Study chargée depuis : " << RESET
             << GREEN << filename 
             << RESET << endl;   
    } else {
        cout << BOLDRED << "Impossible d'ouvrir le fichier " << filename 
             << RESET << endl;
    }
}


void Study :: run() {

    if (compteur == 0) {
        cout << "Equilibrage initial à la température : " << Temperatures[0] << endl;
        montecarlo.equilibrate();
    }

    while(compteur < tempSample){
        cout << endl << BOLDYELLOW << "Temperature " 
             << compteur + 1 << '/' << tempSample << " : " << RESET 
             << YELLOW << Temperatures[compteur] 
             << RESET << endl;

        montecarlo.changeTemperature(Temperatures[compteur]);
        montecarlo.equilibrate();
        montecarlo.calculate();
        montecarlo.meanEnergie(Energies[compteur]);
        montecarlo.meanOrder(Orders[compteur]);

        cout << CYAN << "     Paramètre d'ordre : " << Orders[compteur][0]
             << " +/- " << Orders[compteur][1]
             << RESET << endl;

        cout << CYAN << "     Energies : " << Energies[compteur][0]
             << " +/- " << Energies[compteur][1]
             << RESET << endl;

        compteur++;
        studySave();
    }
}

void Study :: studySave() const{

    //Sauvegarde de la lattice
    montecarlo.lattice.saveLattice(outputFile);
    montecarlo.saveMontecarlo(outputFile);

    //Sauvegarde de l'état de Study
    string filename = outputFile + "_study.save";
    ofstream flux(filename.c_str(), ios::out | ios::trunc);
    if(flux) {
        flux << outputFile << endl;
        flux << tempSample << endl;
        flux << compteur << endl;
        for (int t = 0; t < tempSample ; ++t) {
            flux << fixed << setprecision(10)
                 << Temperatures[t] << ' '
                 << Energies[t][0] << ' '
                 << Energies[t][1] << ' '
                 << Orders[t][0] << ' '
                 << Orders[t][1]
                 << endl;
        }
        cout << GREEN << "     Study enregistrée dans : " << filename.c_str() 
             << RESET << endl;  
    } else {
        cout << BOLDRED << "Impossible d'ouvrir le fichier " << filename 
             << RESET << endl;
    }
}

File: Code/Python/Study.py
from Lattice import Lattice
from MonteCarlo import MonteCarlo

import numpy as np
import matplotlib.pyplot as plt

import pickle

class Study:

    def __init__(self, size, equiSample, meanSample, eRstart, eRend, eRsample, start, reLoad = False, filename='studySave.save'):
        """Constructeur de la classe qui nécessite :
        - la taille de la lattice
        - le nombre de cycle réalisé pour équilibrer le système à chaque température
        un cycle correspond à une série de size**3 changement (1 sur chaque site)
        - le nombre de sample sur lequel on calcule la moyenne (en nombre de cycle)
        - le energieRatio auquel on commence
        - le energieRatio auquel on finit
        - le nombre de energieRatio qu'on veut observer
        - l'information si on part d'une configuration random ou de l'état fondamental
        """

        ##Paramètre de l'étude##
        self.eRstart = eRstart
        self.eRend = eRend
        self.eRsample = eRsample
        self.energieRatios = np.linspace(eRstart, eRend, num=eRsample)
        self.monteCarlo = MonteCarlo(
            size, equiSample, meanSample, self.energieRatios[0], start)
        self.equiSample = equiSample

        #Paramètre de résultats
        self.orderParameter = np.zeros((eRsample, 2))
        self.energy = np.zeros((eRsample, 2))

        #Paramètre d'avancée
        self.i = 0

        #Pour le stockage et le chargement
        self.filename = filename
        if reLoad :
            self.loadClass()

    def run(self):
        """Lance une simulation sur toutes les températures dans energiesRatios"""
        
        if self.i==0 :
            print(
                "Equilibrage inital à la température : {}".format(
                    self.energieRatios[0]))
            
            self.monteCarlo.equilibrate()       

        while self.i < self.eRsample :
            energieRatio = self.energieRatios[self.i]
            print("Température {}/{} : {}".format(
                self.i+1, self.eRsample, energieRatio))
            self.monteCarlo.energieRatio = energieRatio

            self.monteCarlo.equilibrate()
            self.monteCarlo.calculate()

            self.orderParameter[self.i, :] = self.monteCarlo.meanOrderParameter()
            self.energy[self.i, :] = self.monteCarlo.meanEnergy()
            self.i+=1
            self.saveClass()
            print("\n")

    def display(self):
        """Fonction qui affiche l'évolution de l'énergie"""

        plt.plot(self.energieRatios, self.orderParameter[:,1])
        plt.title("Paramètre d'ordre")
        plt.show()

        plt.plot(self.energieRatios, self.energy[:,1])
        plt.title("Energie")
        plt.show()

    def saveClass(self):
        """Fonction qui sauvegarde toute la classe"""

        with open(self.filename,'wb') as f :
            pickle.dump(self.__dict__,f)


    def loadClass(self):
        """Fonction qui recharge toute la classe"""

        with open(self.filename,'rb') as f :
            tmp_dict = pickle.load(f)

        self.__dict__.update(tmp_dict) 
        self.i+=10

if __name__ == '__main__':

    print('Test 3D \n')

    test = Study(30,1000,300,1,1.3,60,'groundstate',True,'10-13.save')
    test.run()